drivers/led/led_pwm.c

```diff
@@ -110,9 +113,89 @@ static int led_pwm_init(const struct device *dev)
 		}
 	}
 
+#ifdef CONFIG_PM_DEVICE
+	config->pm_state = DEVICE_PM_ACTIVE_STATE;
+#endif
+
 	return 0;
 }
 
+#ifdef CONFIG_PM_DEVICE
+
+static int led_pwm_pm_get_state(struct led_pwm_config *config, uint32_t *state)
+{
+	unsigned int key = irq_lock();
+	*state = config->pm_state;
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int led_pwm_pm_set_state(struct led_pwm_config *config, uint32_t new_state)
+{
+	uint32_t old_state;
+	unsigned int key;
+
+	key = irq_lock();
+	old_state = config->pm_state;
+	irq_unlock(key);
+
+	if (old_state == new_state) {
+		/* leave unchanged */
+		return 0;
+	}
+
+	/* switch all underlying PWM devices to the new state */
+	for (size_t i = 0; i < config->num_leds; i++) {
+		const struct led_pwm *led_pwm = &config->led[i];
+
+		LOG_DBG("Switch PWM %p to state %" PRIu32, led_pwm->dev, new_state);
+		int err = device_set_power_state(led_pwm->dev, new_state, NULL, NULL);
+
+		if (err) {
+			LOG_ERR("Cannot switch PWM %p power state", led_pwm->dev);
+		}
+	}
+
+	/* record the new state */
+	key = irq_lock();
+	config->pm_state = new_state;
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int led_pwm_pm_control(const struct device *dev, uint32_t ctrl_command,
+			      void *context, device_pm_cb cb, void *arg)
+{
+	int err;
+
+	LOG_DBG("");
```

Comment:
Why?

Reply:
Since I don't recall logger to add function name it has limited sense.  Will add a function name.

Reply:
answered too quick. it adds function name so I think the original log was fine
Now it looks like below
```
<dbg> led_pwm.led_pwm_pm_control: led_pwm_pm_control
```

Please tell if you prefer to switch to the original.

Reply:
LOG_DBG macro indeed prefixes the function name but I was asking why do need a dummy `LOG_DBG("")` here. The LOG macro below will also have the function name prefixed.

Reply:
Isn't it the only log macro in this function?
there is one when switching happens but this will log message only when state switch happens, there is also no log at get state.

I can remove it if you prefer - I don't mind placing my own logs as I debug - but I though it might be useful if one wish to observe PM related calls

Reply:
removed; can restore it if needed